File: server/shared/utils/gameLogic.ts
import type { Card, Player } from '../types/index.js';

/**
 * Check if a player has busted (duplicate number cards)
 * Returns true if busted, false otherwise
 * 
 * A bust occurs when a player draws a number card that matches
 * a number card they already have. For example:
 * - Player has: 5, 7, 10
 * - Player draws: 7 (BUST! - already has 7)
 * - Player draws: 3 (OK - new unique number)
 * 
 * Note: You CANNOT bust on:
 * - Action cards (FREEZE, FLIP THREE, SECOND CHANCE)
 * - Modifier cards (+2, +4, +6, +8, +10, x2)
 * - The initial card dealt at the start of a round
 */
export function checkBust(player: Player, newCard: Card): boolean {
  if (newCard.type !== 'number' || newCard.value === undefined) {
    return false; // Can't bust on non-number cards
  }

  // Check if player already has this number (before adding the new card)
  // Count how many cards with this value the player already has
  const existingCount = player.numberCards.filter(
    card => card.type === 'number' && card.value === newCard.value
  ).length;

  // If player already has this number, it's a bust
  return existingCount > 0;
}

/**
 * Check if a player has achieved Flip 7 (7 unique number cards)
 */
export function hasFlip7(player: Player): boolean {
  const uniqueNumbers = new Set(
    player.numberCards
      .filter(card => card.type === 'number' && card.value !== undefined)
      .map(card => card.value)
  );
  return uniqueNumbers.size >= 7;
}

/**
 * Calculate player's score for the current round
 */
export function calculateScore(player: Player): number {
  // Step 1: Add the value of Number cards
  let score = player.numberCards.reduce((sum, card) => {
    if (card.type === 'number' && card.value !== undefined) {
      return sum + card.value;
    }
    return sum;
  }, 0);

  // Step 2: If you have the x2 multiplier, double your score
  const x2Card = player.modifierCards.find(
    card => card.modifierType === 'multiply' && card.modifierValue === 2
  );
  
  if (x2Card) {
    score = score * 2;
  }

  // Step 3: Add any additional bonus points (+2, +4, +6, +8, +10)
  const addModifiers = player.modifierCards.filter(
    card => card.modifierType === 'add' && card.modifierValue !== undefined
  );
  
  for (const modifier of addModifiers) {
    if (modifier.modifierValue) {
      score += modifier.modifierValue;
    }
  }

  // Step 4: If you Flip 7 Number cards, score an additional 15 points
  if (hasFlip7(player)) {
    score += 15;
  }

  return score;
}

/**
 * Organize player cards into separate arrays
 */
export function organizePlayerCards(cards: Card[]): {
  numberCards: Card[];
  modifierCards: Card[];
  actionCards: Card[];
} {
  return {
    numberCards: cards.filter(card => card.type === 'number'),
    modifierCards: cards.filter(card => card.type === 'modifier'),
    actionCards: cards.filter(card => card.type === 'action'),
  };
}

/**
 * Get active players (not busted, not stayed)
 */
export function getActivePlayers(players: Player[]): Player[] {
  return players.filter(player => player.isActive && !player.hasBusted);
}

File: app/services/game_logic.rb
# Ported from shared/utils/gameLogic.ts
# Pure functions for game rule calculations

module GameLogic
  # Check if drawing new_card would cause a bust (duplicate number)
  # Cannot bust on: action cards, modifier cards, or initial deal
  def self.bust?(player, new_card)
    return false unless new_card[:type] == "number"

    existing_numbers = player[:number_cards]
      .select { |c| c[:type] == "number" }
      .map { |c| c[:value] }

    existing_numbers.include?(new_card[:value])
  end

  # Check if player has 7 unique number cards (Flip 7)
  def self.flip7?(player)
    unique_numbers = player[:number_cards]
      .select { |c| c[:type] == "number" && !c[:value].nil? }
      .map { |c| c[:value] }
      .uniq

    unique_numbers.size >= 7
  end

  # Calculate player's round score
  # 1. Sum number card values
  # 2. Apply x2 multiplier if present
  # 3. Add modifier bonuses (+2, +4, +6, +8, +10)
  # 4. Add Flip 7 bonus (+15) if applicable
  def self.calculate_score(player)
    # Step 1: Sum number cards
    score = player[:number_cards]
      .select { |c| c[:type] == "number" && !c[:value].nil? }
      .sum { |c| c[:value] }

    # Step 2: x2 multiplier
    has_x2 = player[:modifier_cards].any? { |c| c[:modifier_type] == "multiply" && c[:modifier_value] == 2 }
    score *= 2 if has_x2

    # Step 3: Add modifiers
    player[:modifier_cards]
      .select { |c| c[:modifier_type] == "add" && c[:modifier_value] }
      .each { |c| score += c[:modifier_value] }

    # Step 4: Flip 7 bonus
    score += 15 if flip7?(player)

    score
  end

  # Organize cards into separate arrays by type
  def self.organize_cards(cards)
    {
      number_cards: cards.select { |c| c[:type] == "number" },
      modifier_cards: cards.select { |c| c[:type] == "modifier" },
      action_cards: cards.select { |c| c[:type] == "action" }
    }
  end

  # Get players who are still active (not busted, not stayed/frozen)
  def self.active_players(players)
    players.select { |p| p[:is_active] && !p[:has_busted] }
  end
end
